### ISSUE TITLE
API Limit Exceeded

### ISSUE THREAD

**Issue by amitsehgal** [spinnaker/spinnaker#751]
We are getting "API Limit exceeded" from AWS in clouddriver. On further investigation , we have identified a cause in line: 

https://github.com/spinnaker/clouddriver/blob/master/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ClusterCachingAgent.groovy#L302

When there's no ASG-names provided explicitly, The method listed above is called with "null" argument and tries to fetch all policies and hence throws this error. Temporarily, we have fixed the issue by calling load with list of ASGs and synchronized the caching agent calls to request serially than in parallel. 

However, was wondering why are we getting the error ? Reported the same in slack a while ago.This is preventing us to use spinnaker effectively.

https://spinnakerteam.slack.com/messages/general/search/API%20Rate%20Exceeded/

**Comment by ajordens** [comment]
Hmm.

How large is your footprint, do you have a lot of scaling activities?

Can you share the specific stacktrace from clouddriver ?

**Comment by amitsehgal** [comment]
@ajordens 
I have 150 ASGs running ~2500 instances on 8 VPCs, ~7500 volumes, ~850 ELBs , ~80 policies in a development account.

Here's the stack trace. I had to remove my code fix to get the stack trace. So line numbers might be off compare to current code base. Here's is call

def scalingPolicies = loadScalingPolicies(clients) -----> calls the following method
private Map<String, List<Map>> loadScalingPolicies(AmazonClients clients) {
    loadScalingPolicies(clients, null) //the culprit
}


com.amazonaws.AmazonServiceException: CloudWatchAlarm Rate exceeded (Service: AmazonAutoScaling; Status Code: 400; Error Code: Throttling; Request ID: a217a0a2-da7e-11e5-962a-a1917861e2d6)
at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:1160)
at com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:748)
at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:467)
at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:302)
at com.amazonaws.services.autoscaling.AmazonAutoScalingClient.invoke(AmazonAutoScalingClient.java:3246) at com.amazonaws.services.autoscaling.AmazonAutoScalingClient.describePolicies(AmazonAutoScalingClient.java:511)
at com.amazonaws.services.autoscaling.AmazonAutoScaling$describePolicies$1.call(Unknown Source)
at com.netflix.spinnaker.oort.aws.provider.agent.ClusterCachingAgent.loadScalingPolicies(ClusterCachingAgent.groovy:309)
at com.netflix.spinnaker.oort.aws.provider.agent.ClusterCachingAgent.this$2$loadScalingPolicies(ClusterCachingAgent.groovy)
at com.netflix.spinnaker.oort.aws.provider.agent.ClusterCachingAgent$this$2$loadScalingPolicies$1.callCurrent(Unknown Source)
at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:133)
at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:145)
at com.netflix.spinnaker.oort.aws.provider.agent.ClusterCachingAgent.loadScalingPolicies(ClusterCachingAgent.groovy:297)
at com.netflix.spinnaker.oort.aws.provider.agent.ClusterCachingAgent.this$2$loadScalingPolicies(ClusterCachingAgent.groovy)
at com.netflix.spinnaker.oort.aws.provider.agent.ClusterCachingAgent$this$2$loadScalingPolicies$0.callCurrent(Unknown Source)
at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:133)
at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:141)
at com.netflix.spinnaker.oort.aws.provider.agent.ClusterCachingAgent.loadData(ClusterCachingAgent.groovy:389)
at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgent(CachingAgent.java:63)
at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentExecutionAction.execute(ClusteredAgentScheduler.java:189)
at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentJob.run(ClusteredAgentScheduler.java:164)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)

**Comment by ravi-dubey** [comment]
Beuhler?

**Comment by medyagh** [comment]
we are seeing this too.

**Comment by ajordens** [comment]
Same or different limit being exceeded @medyagh?

**Comment by medyagh** [comment]
@ajordens 
CloudDriver log says 
```
CloudWatchAlarm Rate exceeded (Service: AmazonAutoScaling; Status Code: 400; Error Code: Throttling; Request ID: 41936410-08b5-11e6-be0b-e1281f8ffe86)
    at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:1389)
    at com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:902)
    at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:607)
    at com.amazonaws.http.AmazonHttpClient.doExecute(AmazonHttpClient.java:376)
    at com.amazonaws.http.AmazonHttpClient.executeWithTimer(AmazonHttpClient.java:338)
    at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:287)
    at com.amazonaws.services.autoscaling.AmazonAutoScalingClient.invoke(AmazonAutoScalingClient.java:3046)
    at com.amazonaws.services.autoscaling.AmazonAutoScalingClient.describePolicies(AmazonAutoScalingClient.java:1600)
    at com.amazonaws.services.autoscaling.AmazonAutoScaling$describePolicies$1.call(Unknown Source)
    at com.netflix.spinnaker.clouddriver.aws.provider.agent.ClusterCachingAgent.loadScalingPolicies(ClusterCachingAgent.groovy:305)
    at com.netflix.spinnaker.clouddriver.aws.provider.agent.ClusterCachingAgent.this$2$loadScalingPolicies(ClusterCachingAgent.groovy)
    at com.netflix.spinnaker.clouddriver.aws.provider.agent.ClusterCachingAgent$this$2$loadScalingPolicies$1.callCurrent(Unknown Source)
    at com.netflix.spinnaker.clouddriver.aws.provider.agent.ClusterCachingAgent.loadScalingPolicies(ClusterCachingAgent.groovy:293)
    at com.netflix.spinnaker.clouddriver.aws.provider.agent.ClusterCachingAgent.this$2$loadScalingPolicies(ClusterCachingAgent.groovy)
    at com.netflix.spinnaker.clouddriver.aws.provider.agent.ClusterCachingAgent$this$2$loadScalingPolicies$0.callCurrent(Unknown Source)
    at com.netflix.spinnaker.clouddriver.aws.provider.agent.ClusterCachingAgent.loadData(ClusterCachingAgent.groovy:385)
    at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgent(CachingAgent.java:63)
    at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentExecutionAction.execute(ClusteredAgentScheduler.java:195)
    at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentJob.run(ClusteredAgentScheduler.java:169)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```
and I tried changing the redis.poll.intervalSeconds`  and redis.parallelism that didn't help. I also upgraded the clouddriver.

**Comment by bkarakashev** [comment]
Hey @ajordens  @medyagh In which file I need to do this Redis changes (**redis.poll.intervalSeconds**, **redis.parallelism**, **redis.scheduler**) that you suggested. Is it some of the config files for Spinnaker?

**Comment by riltsken** [comment]
Wanted to chime in and say that upping the redis poll interval didn't really help us. What we did was contact AWS support and specifically asked to increase the api throttle limit for the above call replicated via cli `aws autoscaling --region=us-east-1 describe-policies` and everything is good now. We went the route of asking through the autoscaling team since it goes through their api but I believe they just upped our cloudwatch limits.

**Comment by ravi-dubey** [comment]
@riltsken, thanks for the feedback.  We are exploring moving to a push-pull model to keep the cache up to date without polling/brute force reloading.  We hit API limits as an organization when multiple kinds of cloud tooling (primarily the AWS console) are invoked in a short timeframe by hundreds of developers.  Spinnaker is perceived as contributing to the problem and we want to reverse that perception.

**Comment by cdoughty-r7** [comment]
We are actually seeing this in our account as well.  Was there ever a resolution or suggested work around to this issue?

**Comment by cfieber** [comment]
There are a couple of configuration options that might help you out:

1) use serviceLimits to rate limit calls. If you are seeing throttling on a particular API endpoint then you can limit that specifically. Right now a rate limit of 10 requests per second to each amazon service in each region is the default, but that is likely too aggressive for some endpoints such as cloudwatch. See spinnaker/clouddriver#1291 for an example of the configuration options there

2) increase the number of retries for the AWS SDK. Right now the SDK will retry with exponential backoff but the default number of retries is 3. Set `aws.client.maxErrorRetry` to a higher number than that in your clouddriver configuration.

**Comment by bcall-r7** [comment]
Just to confirm what @riltsken stated. We went the same route and and requested a limit increase for CloudWatch DescribeAlarms which was where we were getting limited when describe-policies was being called. Amazon increased our limit significantly after proving information on the amount of requests being made. This solved our problem with the CloudWatch rate limit however we are still receiving AmazonAutoScalingException exceptions due to another API rate limit we have yet to solve.

**Comment by lwander** [comment]
Closing since workarounds have been documented https://github.com/spinnaker/spinnaker/issues/751#issuecomment-305584887

**Comment by valter-silva-git** [comment]
Hi everyone, just a quick update: I'm facing this issue again, the issue seems to appear on versions 1.8.5 and 1.9.0. So only rolling-back to version 1.8.1 seems to fix the issue. Any suggestions?

**Comment by 7oas7er** [comment]
I think the issue is not related to the above mentioned limits @valter-silva-au, but rather to hitting an API limit at AWS.
With Spinnaker 1.9 I get this:

Exception ( Monitor Deploy )
Rate exceeded (Service: AmazonAutoScaling; Status Code: 400; Error Code: Throttling; Request ID: ec3e8bf4-c987-11e8-afa0-23ba773e2599)

I will try to set the clouddriver API rate limits as described here though I do not now what an appropriate value could be: https://blog.armory.io/fine-grained-rate-limits-for-spinnaker-clouddriver/
